### ISSUE TITLE
Unable to read char* fields in objects

### ISSUE THREAD

**Issue by wdconinc** [scikit-hep/uproot#31]
It seems that the remaining issue in #28 was caused by a char* field in an object. Example ROOT file and code attached. The issue is with the streamer for the `name` field in the `mydata` class.

[test.root.gz](https://github.com/scikit-hep/uproot/files/1533311/test.root.gz)

```
#include "TFile.h"
#include "TTree.h"

class mydata : public TObject {
    public:
        mydata() { };
        virtual ~mydata() { };

        int size;
        char* name;
        void set(const char* n, int s) {
                size = s;
                name = new char[s];
                strncpy(name,n,s);
        }

        ClassDef(mydata, 1);
};

void test_uproot()
{
  TFile file("test.root","RECREATE");
  TTree tree("T","T");

  mydata data;

  tree.Branch("data",&data);

  for (int i = 0; i < 1; i++) {
    data.set("test",4);
    tree.Fill();
  }

  tree.Write();

  file.Write();
  file.Close();
}
```

```
In [1]: import uproot
In [2]: uproot.open("test.root")
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-2-295bb97bcaec> in <module>()
----> 1 uproot.open("test.root")

/usr/local/lib/python3.5/dist-packages/uproot-2.5.2-py3.5.egg/uproot/rootio.py in open(path, localsource, xrootdsource, **options)
     57     if _bytesid(parsed.scheme) == b"file" or len(parsed.scheme) == 0:
     58         path = parsed.netloc + parsed.path
---> 59         return ROOTDirectory.read(localsource(path), **options)
     60 
     61     elif _bytesid(parsed.scheme) == b"root":

/usr/local/lib/python3.5/dist-packages/uproot-2.5.2-py3.5.egg/uproot/rootio.py in read(source, *args, **options)
    138                     streamerinfos, streamerinfosmap, streamerrules = [], {}, []
    139 
--> 140                 classes = _defineclasses(streamerinfos)
    141                 context = ROOTDirectory._FileContext(source.path, streamerinfos, streamerinfosmap, classes, uproot.source.compressed.Compression(fCompress), tfile)
    142 

/usr/local/lib/python3.5/dist-packages/uproot-2.5.2-py3.5.egg/uproot/rootio.py in _defineclasses(streamerinfos)
    625                         basicnames.append("self.{0}".format(_safename(element.fName)))
    626                         fields.append(_safename(element.fName))
--> 627                         basicletters += _ftype2struct(element.fType)
    628 
    629                         if elementi + 1 == len(streamerinfo.fElements) or not isinstance(streamerinfo.fElements[elementi + 1], TStreamerBasicType) or streamerinfo.fElements[elementi + 1].fArrayLength != 0:

/usr/local/lib/python3.5/dist-packages/uproot-2.5.2-py3.5.egg/uproot/rootio.py in _ftype2struct(fType)
    566         return "d"
    567     else:
--> 568         raise NotImplementedError(fType)
    569 
    570 def _safename(name):

NotImplementedError: 7
```

**Comment by jpivarski** [comment]
Thanks! I'm surprised that I missed that one because we have some other string types— I just have to map it to the correct container. I'll take care of it as soon as I can.

**Comment by jpivarski** [comment]
It turned out to be more than just linking up a data type with an interpretation. I hadn't realized that the NotImplementedError was during the streamer reading— I've been trying to keep the streamer reading free of NotImplementedErrors because it happens automatically when loading a file; failing to load a file because I can't recognize some data type in it that the user might not want to read anyway is a bad surprise. The exception should happen when the user tries to read it, indicating that the user really wants data of this type. Most of the NotImplementedErrors have already been put into the _generated_ code, rather than the _generating_ code, but not this one. (The one remaining in `_ftype2struct` is unavoidable— I _have to_ know how many bytes to put in that Struct, and without guidance from real ROOT files, I just don't know.)

After that, it seems that `char*` needs to be read differently from other strings. Each `char*` has a 4-byte size header, whereas other string types have 1-byte followed by 4-bytes if the first byte is 255. Thus, `char*` is more regular but less compact than other string types.

It's also interesting that the ROOT file said this string has six bytes. The per-string header and the end-of-baskets offsets are in agreement on that: the string is `"test\xed\x7f"`, not `"test"`. Looking at this in C++ ROOT, I see the same thing. Extra bytes are included after the desired string.

Ah! Figured it out! The problem was in writing the ROOT file. Your call to `strncpy` did not include the termination byte, so ROOT didn't know where to stop. (ROOT doesn't know that your `size` field corresponds to the size of your `name` field.) You probably had `"\xed\x7f\x00"` as random gunk in memory, and ROOT copied that. That's actually a security flaw: there could have been passwords in memory before reaching the `"\x00"`. Not in this case, though.

I changed your generating script to make a new `issue31.root` (thanks for including it!):

```c++
  data.set("one",3+1);
  tree.Fill();

  data.set("two",3+1);
  tree.Fill();

  data.set("three",5+1);
  tree.Fill();

  data.set("four",4+1);
  tree.Fill();

  data.set("five",4+1);
  tree.Fill();
```

And now when we read `uproot.open("tests/samples/issue31.root")["T"].array("name").tolist()`, we get

```
[b"one", b"two", b"three", b"four", b"five"]
```
